src/platform/backends/lxd/lxd_virtual_machine.cpp

```diff
@@ -0,0 +1,386 @@
+            config["user.meta-data"] = QString::fromStdString(mpu::emit_cloud_config(desc.meta_data_config));
+
+        if (!desc.vendor_data_config.IsNull())
+            config["user.vendor-data"] = QString::fromStdString(mpu::emit_cloud_config(desc.vendor_data_config));
+
+        if (!desc.user_data_config.IsNull())
+            config["user.user-data"] = QString::fromStdString(mpu::emit_cloud_config(desc.user_data_config));
+
+        QJsonObject devices{
+            {"config", QJsonObject{{"source", "cloud-init:config"}, {"type", "disk"}}},
+            {"root", QJsonObject{{"path", "/"},
+                                 {"pool", "default"},
+                                 {"size", QString::number(get_minimum_disk_size(desc.disk_space).in_bytes())},
+                                 {"type", "disk"}}}};
+
+        QJsonObject virtual_machine{
+            {"name", name},
+            {"config", config},
+            {"devices", devices},
+            {"source", QJsonObject{{"type", "image"},
+                                   {"mode", "pull"},
+                                   {"server", QString::fromStdString(desc.image.stream_location)},
+                                   {"protocol", "simplestreams"},
+                                   {"fingerprint", QString::fromStdString(desc.image.id)}}}};
+
+        auto json_reply = lxd_request(manager, "POST", QUrl(QString("%1/virtual-machines").arg(base_url.toString())),
+                                      virtual_machine);
+        mpl::log(mpl::Level::debug, name.toStdString(),
+                 fmt::format("Got LXD creation reply: {}", QJsonDocument(json_reply).toJson()));
+
+        if (json_reply["metadata"].toObject()["class"] == QStringLiteral("task") &&
+            json_reply["status_code"].toInt(-1) == 100)
+        {
+            QUrl task_url(QString("%1/operations/%2")
+                              .arg(base_url.toString())
+                              .arg(json_reply["metadata"].toObject()["id"].toString()));
+
+            // Instead of polling, need to use websockets to get events
+            while (true)
+            {
+                try
+                {
+                    auto task_reply = lxd_request(manager, "GET", task_url);
+
+                    if (task_reply["error_code"].toInt(-1) != 0)
+                    {
+                        break;
+                    }
+
+                    auto status_code = task_reply["metadata"].toObject()["status_code"].toInt(-1);
+                    if (status_code == 200)
+                    {
+                        break;
+                    }
+                    else
+                    {
+                        std::this_thread::sleep_for(5s);
+                    }
+                }
+                // Implies the task is finished
+                catch (const LXDNotFoundException& e)
+                {
+                    break;
+                }
+            }
+        }
```

Comment:
Should this go into lxd_request?

Reply:
I kind of think the caller of `lxd_request()` should determine how it waits on an operation to finish.  Some callers may use the `operation/blah/wait` method while others may use a polling (and later websockets) method.

Reply:
Sure, it would have to be the caller's decision whether to let `lxd_request()` handle the async operation or not. I'm kinda thinking `::waitForFinished()` style. Except it's not an object, I know ;)

Reply:
Well, still, the caller should determine _how_ it waits on an operation to complete.  I suppose we could have a default wait logic in `lxd_request` and if the caller doesn't want or need to use that, then it can pass in `0` for the timeout and `lxd_request` will just return the json response from LXD.  WDYT?

Reply:
That's roughly what I had in mind. Maybe not inferred from `timeout == 0` but rather something explicit like `bool wait`. Ideally something that resembles something already existing :)

Reply:
Hmm, ok, I'll basically put it back to how you originally had it :grin: